### PR TITLE
Delayed import

### PR DESCRIPTION
**Pull request by denismakogon**
# What's being done?
Customer's code import happens right before the first calls.

# Improvement
With this change, the cold start only affected by FDK itself,
but not by the customer's code.

# Minimal python version
3.7.1

Based on #61

### CODE REVIEW

**Comment by rdallman** [comment]
gonna rebase, if you're around ping me and I won't

**Comment by rdallman** [comment]
https://github.com/fnproject/fdk-python/tree/delayed-import-feature-original original is here if rebase messed up, but git diff says +/- 0 -- viewing diff now...

**Comment by denismakogon** [comment]
Review order:
1. #61 
2. #62 
3. #60

**Comment by denismakogon** [comment]
> are we dropping support for python3.6? is there a reason? do we know if this effects many users? python3.7 came out not even 6 months ago, python3.6 is 2 years old at least, is this reasonable to expect if 3.7 is required now?

`importutil` lib is available in py3.7 with its delayed imports feature.

> does this break existing user functions in any way? or is this a compatible feature that they may use if they want to update their code? or maybe they don't need to update? asking about compatibility in general if those questions are too specific. asking because the examples here are changed, and don't necessarily want to break users if we can help it. it also appears like the body we're handing off we're not deserializing to a typed object anymore, i think that's probably okay, just would like to understand usage here for any breakages, since they are not enumerated anywhere I can see...

yes, data types are changing, specifically `data` in function's definition. regarding changes in general -i'm working on medium post about the details of a newer Python FDK, so every major change would be outlined.

File: setup.cfg [hunk 1]
```diff
@@ -23,3 +23,7 @@ packages =
 [global]
 setup-hooks =
     pbr.hooks.setup_hook
+
+[entry_points]
+console_scripts =
+    fdk = fdk.scripts.main:main
```
Comment (denismakogon) [inline setup.cfg]: this should be:
```
fdk = fdk.scripts.fdk:main

File: requirements.txt [hunk 2]
```diff
@@ -1,6 +1,7 @@
 pbr!=2.1.0,>=2.0.0 # Apache-2.0
 uvloop==0.11.3
 ujson==1.35
+requests==2.20.1
```
Comment (denismakogon) [inline requirements.txt]: shouldn't be here, please delete

**Comment by rdallman** [comment]
> yes, data types are changing, specifically data in function's definition. regarding changes in general -i'm working on medium post about the details of a newer Python FDK, so every major change would be outlined.

ok. i know we have at least one other user, keeping compatibility might be nice even if we deprecate it, gives some time to move over without breaking. I'm thinking about function definitions too (in their actual code) because of the delayed imports, it seems like no more 'main' at least from the example. I'll wait for the blog post, thanks for writing it, would just advise to keep required migrations to a minimum, if possible (not always possible). would like to get both of these patches in this week

**Comment by denismakogon** [comment]
> i know we have at least one other user, keeping compatibility might be nice even if we deprecate it, gives some time to move over without breaking

That's quite interesting, because deploying a new function means that you upgrading it, a version going to be bumped. So, it's up to us to notify whether we changing the API and it's up to a developers to check and test the compatibility, that's why FDK provides a framework do to a unit tests and that's why an Python FDK docs explicitly telling to developers that CLI generates you a very basic boilerplate, but you need to add a unit tests by yourself. So, the best we can do and we do for now is keeping the signature of a user's function as it is.

>  it seems like no more 'main' at least from the example

Main has gone because user's code no longer considered as an entry-point to a function, because user's module initialization going to be delayed, that's the whole idea of this PR.

> would just advise to keep required migrations to a minimum

that's the biggest change of all time that happens in Python FDK, at least until we introduce a new triggers.

**Comment by rdallman** [comment]
> So, it's up to us to notify whether we changing the API and it's up to a developers to check and test the compatibility, 

disagree - it's up to us to maintain a stable API at some point (soon, if not already) too. this misses the point, I'm simply asking if it's possible to keep compatibility while adding this new feature, without a user that upgrades the fdk having to change their code (this was the reason for #63 - asyncio/h11 changes do not need to break user code when upgrading on their own as it's all behind the scenes, and #63 is all that we need to unblock the current issues we're having with aiohttp... delayed import is a complementary change, that we don't necessarily need, and will break user code much more it seems...)

> Main has gone because user's code no longer considered as an entry-point to a function, because user's module initialization going to be delayed, that's the whole idea of this PR.

I guess i'm asking as a user that has a python func that uses main, what happens when i try to update the fdk to the version with delayed imports without changing the code? changes are required? is it possible not to require this and have both modes of operation possible from the newer fdk?

**Comment by denismakogon** [comment]
> I'm simply asking if it's possible to keep compatibility while adding this new feature

No, it's impossible, users need to change their code and change their dockerfiles (if they use custom ones).

> delayed import is a complementary change, that we don't necessarily need

I strongly disagree with this, it's not a complimentary change, it's almost the most valuable change in FDK so far. Try to write a function that uses tensorflow with keras and mathplotlib you'll notice that a function will take almost ~40 seconds to start. So, in order to reduce cold start problem we MUST, i repeat, we MUST, solve this problem in two ways:
 - get rid of aiohttp
 - make FDK do a delayed imports of customer's code

without that changing aiohttp to asyncio + h11 is pointless and meaningless, because you attempting to cure the symptoms, but not a disease.

**Comment by rdallman** [comment]
I agree that delayed import is useful and we should add it, however our problem as it stands is aiohttp load time on its own (and we need to import an actual http library to handle the 5s timeout we're hitting without being able to delay it), and a user is blocked on this, and the delayed import change is not required to unblock this specific case... i appreciate your understanding of the issue, i'm just thinking about the time to land the delayed import change, write the blog post, help users migrate their code, fix any issues, and it's a lot longer tail to landing the fix for the specific thing blocking us right now (asyncio/h11) while also being an additional work stream (asyncio/h11 is useful on its own, and also useful to allow doing delayed import, at least, but in my mind the order is flipped priority wise and dependency wise I guess...). anyway, that's where i'm at, not trying to rule out anything really it seems like that was the interpretation...

**Comment by denismakogon** [comment]
> however our problem as it stands is aiohttp load time on its own 

Once again, you are not solving a problem at all without delayed import. 

> delayed import change is not required to unblock this specific case

Again, it **is required** even more than aiohttp replacement. You're trying to solve a problem without getting into a details of the problem, the biggest issue is customer's code, when they import and FDK and start a handler this is where the problem sits: a customer's code is a problem, but not an FDK itself, as long as customer's code decided whether this (`func.py`) is an entry-point to a function we **CAN'T** solve a problem.

Migrating an FDK to asyncio + h11 DOESN'T solve a problem, period. An entry-point to a function MUST be handled by an FDK, but not a customer's code.

It your concern is a migration process than i ensure you it takes less than 2 minutes per a function. My blogpost will help with that.

Delayed import IS a necessary change, it MUST be merged in order to secure an FDK from blowing up on start because customer code can do lots of various unexpected things that will make an FDK a source of a problem over and over again.

**Comment by rdallman** [comment]
> Delayed import IS a necessary change, it MUST be merged in order to secure an FDK from blowing up on start because customer code can do lots of various unexpected things that will make an FDK a source of a problem over and over again.

i agree about this, but i can see this discussion isn't really going anywhere. i'll reiterate that our problem today as it stands is that running hello world (no dependencies) with the python fdk takes more than 5 seconds on a constrained environment, and fixing that is highest priority...

**Comment by denismakogon** [comment]
> problem today as it stands is that running hello world (no dependencies) with the python fdk takes more than 5 seconds on a constrained environment, and fixing that is highest priority

That’s what’s these PRs are all about, they can’t go separately because they solve one problem altogether from all possible perspectives: from an FDK and customer’s code.

**Pull request closed without merging**